File: modules/offline-core-android/src/main/kotlin/cc/musicme/offline/StreamSession.kt
package cc.musicme.offline

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.Base64

data class TrackRef(val cb: Long, val disc: Int, val track: Int)

data class SessionMetrics(
    val bootstrapMs: Double?,
    val firstRangeMs: Double?,
    val firstDecryptMs: Double?,
    val fileSizeBytes: Long?,
    val sessionRotations: Int,
)

class StreamSession(
    private val workerUrl: String,
    private val tokenProvider: suspend () -> String,
    private val trackRef: TrackRef,
    private val client: OkHttpClient = OkHttpClient(),
) {
    data class Bootstrap(
        val sessionId: String,
        val fileSize: Long,
        val key: ByteArray,
        val iv: ByteArray,
        val contentType: String,
    )

    private val mutex = Mutex()
    @Volatile private var cached: Bootstrap? = null

    @Volatile private var bootstrapMs: Double? = null
    @Volatile private var firstRangeMs: Double? = null
    @Volatile private var firstDecryptMs: Double? = null
    @Volatile private var fileSizeBytes: Long? = null
    private val sessionRotations = java.util.concurrent.atomic.AtomicInteger(0)

    fun metrics(): SessionMetrics = SessionMetrics(
        bootstrapMs, firstRangeMs, firstDecryptMs, fileSizeBytes, sessionRotations.get()
    )

    suspend fun bootstrap(): Bootstrap {
        cached?.let { return it }
        // Single-flight: only one bootstrap at a time. Concurrent callers wait on the
        // mutex then read the fresh cache (re-check inside the lock). Prevents the
        // double-bootstrap that leaked an /init-stream session server-side.
        return mutex.withLock {
            cached ?: runBootstrap(allowRetry = true).also { cached = it }
        }
    }

    suspend fun read(start: Long, endExclusive: Long): ByteArray {
        val boot = bootstrap()
        return try {
            runRead(start, endExclusive, boot)
        } catch (e: OfflineError.StreamRangeFailed) {
            if (e.status == 410) {
                sessionRotations.incrementAndGet()
                mutex.withLock { cached = null }
                val fresh = bootstrap()
                runRead(start, endExclusive, fresh)
            } else throw e
        }
    }

    suspend fun heartbeat(): Boolean {
        val boot = mutex.withLock { cached } ?: return false
        val url = joinUrl(workerUrl, "heartbeat/${boot.sessionId}")
        val req = Request.Builder().url(url)
            .post("{\"duration_ms\":0}".toRequestBody())
            .header("Content-Type", "application/json")
            .build()
        return withContext(Dispatchers.IO) {
            val call = client.newCall(req)
            try {
                call.execute().use { r ->
                    when (r.code) {
                        200 -> true
                        410 -> false
                        else -> throw OfflineError.StreamRangeFailed(r.code)
                    }
                }
            } finally {
                call.cancel()
            }
        }
    }

    fun close() {
        // Synchronous nuke of cache. Mutex is fine to skip — losing a torn
        // read here at worst forces an extra bootstrap on the next op.
        cached = null
    }

    // --- private ---

    private suspend fun runBootstrap(allowRetry: Boolean): Bootstrap {
        val t0 = System.nanoTime()
        val token = tokenProvider()
        val body = """{"cb":${trackRef.cb},"disc":${trackRef.disc},"track":${trackRef.track}}"""
        val req = Request.Builder()
            .url(joinUrl(workerUrl, "init-stream"))
            .post(body.toRequestBody())
            .header("Authorization", "Bearer $token")
            .header("Content-Type", "application/json")
            .build()
        val boot = withContext(Dispatchers.IO) {
            val call = client.newCall(req)
            try {
                call.execute().use { r ->
                    when (r.code) {
                        401 -> if (allowRetry) runBootstrapInner() else throw OfflineError.SessionUnauthorized
                        200 -> parseBootstrap(r.body!!.string())
                        else -> throw OfflineError.SessionInitFailed(r.code)
                    }
                }
            } finally {
                call.cancel()
            }
        }
        // Record only on the outermost successful call (allowRetry == true means top-level).
        // The recursive inner call (allowRetry == false) also records if it succeeds,
        // but "first write wins" via the null-check.
        if (bootstrapMs == null) bootstrapMs = (System.nanoTime() - t0) / 1_000_000.0
        fileSizeBytes = boot.fileSize
        return boot
    }
    private suspend fun runBootstrapInner(): Bootstrap = runBootstrap(allowRetry = false)

    private fun parseBootstrap(text: String): Bootstrap {
        try {
            fun str(key: String): String =
                Regex(""""$key"\s*:\s*"([^"]+)"""").find(text)?.groupValues?.get(1)
                    ?: error("missing $key")
            fun long(key: String): Long =
                Regex(""""$key"\s*:\s*(\d+)""").find(text)?.groupValues?.get(1)?.toLong()
                    ?: error("missing $key")
            return Bootstrap(
                sessionId = str("sessionId"),
                fileSize = long("fileSize"),
                key = Base64.getDecoder().decode(str("keyB64")),
                iv = Base64.getDecoder().decode(str("ivB64")),
                contentType = str("contentType"),
            )
        } catch (e: Exception) {
            throw OfflineError.StreamMalformedResponse("init-stream")
        }
    }

    private suspend fun runRead(start: Long, endExclusive: Long, boot: Bootstrap): ByteArray {
        val aligned = (start / 16) * 16
        val endInclusive = endExclusive - 1
        val url = joinUrl(workerUrl, "stream/${boot.sessionId}")

        var lastError: Throwable? = null
        for (attempt in 0..2) {
            try {
                val tStart = System.nanoTime()
                val req = Request.Builder().url(url)
                    .header("Range", "bytes=$aligned-$endInclusive")
                    .build()
                return withContext(Dispatchers.IO) {
                    val call = client.newCall(req)
                    try {
                        call.execute().use { r ->
                            when {
                                r.code == 410 -> throw OfflineError.StreamRangeFailed(410)
                                r.code == 403 -> throw OfflineError.SessionFingerprintMismatch
                                r.code == 429 || r.code >= 500 -> throw OfflineError.StreamRangeFailed(r.code)
                                r.code !in setOf(200, 206) -> throw OfflineError.StreamRangeFailed(r.code)
                                else -> {
                                    val cipher = r.body!!.bytes()
                                    if (firstRangeMs == null) firstRangeMs = (System.nanoTime() - tStart) / 1_000_000.0
                                    // Always request from `aligned` (16-aligned): the server serves
                                    // [aligned..endInclusive], so the decrypted plaintext maps to that
                                    // same range. To return [start..endExclusive) we must drop the first
                                    // `start-aligned` bytes. The server's X-Skip-Bytes header is always 0
                                    // (it serves from the requested range), so we compute the skip
                                    // locally. The CTR counter is aligned/16 (the 16-byte block number
                                    // containing `aligned`). The web SDK avoided the bug by only ever
                                    // requesting 16-aligned starts (256 KiB chunks).
                                    val skip = (start - aligned).toInt()
                                    val counter = (aligned / 16).toInt()
                                    val tDec = System.nanoTime()
                                    val plain = AESCTRDecryptor.decrypt(cipher, boot.key, boot.iv, counter)
                                    if (firstDecryptMs == null) firstDecryptMs = (System.nanoTime() - tDec) / 1_000_000.0
                                    val wanted = (endExclusive - start).toInt()
                                    val to = minOf(skip + wanted, plain.size)
                                    plain.copyOfRange(skip, to)
                                }
                            }
                        }
                    } finally {
                        call.cancel()
                    }
                }
            } catch (e: OfflineError.StreamRangeFailed) {
                if (e.status == 410 || e.status == 403) throw e
                lastError = e
                delay(listOf(500L, 1_000L, 2_000L)[attempt])
            } catch (e: OfflineError.SessionFingerprintMismatch) {
                throw e
            } catch (e: Exception) {
                lastError = e
                delay(listOf(500L, 1_000L, 2_000L)[attempt])
            }
        }
        throw OfflineError.StreamNetworkExhausted(lastError?.message ?: "unknown")
    }

    private fun joinUrl(base: String, path: String): String {
        val b = base.trimEnd('/')
        val p = path.trimStart('/')
        return "$b/$p"
    }
}

File: demos/react-native/modules/offline/ios/OfflineCore/StreamSession.swift
import Foundation

public final class StreamSession: Sendable {
    public struct TrackRef: Equatable, Sendable {
        public let cb: Int
        public let disc: Int
        public let track: Int
        public init(cb: Int, disc: Int, track: Int) {
            self.cb = cb; self.disc = disc; self.track = track
        }
    }

    public struct Bootstrap: Sendable {
        public let sessionId: String
        public let fileSize: Int64
        public let key: Data
        public let iv: Data
        public let contentType: String
    }

    public struct SessionMetrics: Sendable {
        public let bootstrapMs: Double?
        public let firstRangeMs: Double?
        public let firstDecryptMs: Double?
        public let fileSizeBytes: Int64?
        public let sessionRotations: Int
    }

    private let workerUrl: URL
    private let tokenProvider: @Sendable () async throws -> String
    private let trackRef: TrackRef
    private let urlSession: URLSession
    private let cache = BootstrapCache()
    private let metricsRec = MetricsRecorder()

    public init(workerUrl: URL,
                tokenProvider: @escaping @Sendable () async throws -> String,
                trackRef: TrackRef,
                urlSession: URLSession = .shared) {
        self.workerUrl = workerUrl
        self.tokenProvider = tokenProvider
        self.trackRef = trackRef
        self.urlSession = urlSession
    }

    public func bootstrap() async throws -> Bootstrap {
        if let b = await cache.get() { return b }
        let b = try await runBootstrap(allowRetry: true)
        await cache.set(b)
        return b
    }

    public func read(range: Range<Int64>) async throws -> Data {
        let boot = try await bootstrap()
        do {
            return try await runRead(range: range, boot: boot)
        } catch OfflineError.streamRangeFailed(let status) where status == 410 {
            await metricsRec.incRotation()
            await cache.clear()
            let fresh = try await bootstrap()
            return try await runRead(range: range, boot: fresh)
        }
    }

    public func heartbeat() async throws -> Bool {
        guard let boot = await cache.get() else { return false }
        let url = workerUrl.appendingPathComponent("heartbeat/\(boot.sessionId)")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = "{\"duration_ms\":0}".data(using: .utf8)
        let (_, resp) = try await urlSession.data(for: req)
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        if status == 200 { return true }
        if status == 410 { return false }
        throw OfflineError.streamRangeFailed(status: status)
    }

    public func close() async {
        await cache.clear()
    }

    public func metrics() async -> SessionMetrics {
        await metricsRec.snapshot()
    }

    // MARK: - Private

    private func runBootstrap(allowRetry: Bool) async throws -> Bootstrap {
        let t0 = Date()
        let url = workerUrl.appendingPathComponent("init-stream")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let token = try await tokenProvider()
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        // `context` declares the listening mode for royalty declaratifs and is
        // required by the integration contract (offline downloads are on-demand).
        let body: [String: Any] = [
            "cb": trackRef.cb, "disc": trackRef.disc, "track": trackRef.track,
            "context": "on_demand",
        ]
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, resp) = try await urlSession.data(for: req)
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        if status == 401 {
            if allowRetry { return try await runBootstrap(allowRetry: false) }
            throw OfflineError.sessionUnauthorized
        }
        guard status == 200 else { throw OfflineError.sessionInitFailed(status: status) }
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let sid = json["sessionId"] as? String,
              let fileSize = (json["fileSize"] as? NSNumber)?.int64Value,
              let contentType = json["contentType"] as? String,
              let keyB64 = json["keyB64"] as? String,
              let ivB64 = json["ivB64"] as? String,
              let key = Data(base64Encoded: keyB64),
              let iv = Data(base64Encoded: ivB64)
        else { throw OfflineError.streamMalformedResponse("init-stream") }
        let boot = Bootstrap(sessionId: sid, fileSize: fileSize, key: key, iv: iv, contentType: contentType)
        await metricsRec.recordBootstrap(Date().timeIntervalSince(t0) * 1000.0)
        await metricsRec.recordFileSize(boot.fileSize)
        return boot
    }

    private func runRead(range: Range<Int64>, boot: Bootstrap) async throws -> Data {
        let aligned = (range.lowerBound / 16) * 16
        let endInclusive = range.upperBound - 1
        let url = workerUrl.appendingPathComponent("stream/\(boot.sessionId)")
        var req = URLRequest(url: url)
        req.setValue("bytes=\(aligned)-\(endInclusive)", forHTTPHeaderField: "Range")

        var lastError: Error?
        for attempt in 0..<3 {
            do {
                let tStart = Date()
                let (cipher, resp) = try await urlSession.data(for: req)
                let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
                if status == 410 { throw OfflineError.streamRangeFailed(status: 410) }
                if status == 403 { throw OfflineError.sessionFingerprintMismatch }
                if status >= 500 || status == 429 {
                    throw OfflineError.streamRangeFailed(status: status)
                }
                guard status == 200 || status == 206 else {
                    throw OfflineError.streamRangeFailed(status: status)
                }
                let http = resp as! HTTPURLResponse
                let skip = Int(http.value(forHTTPHeaderField: "X-Skip-Bytes") ?? "0") ?? 0
                let counterStart = Int(http.value(forHTTPHeaderField: "X-Counter-Start") ?? "0") ?? Int(aligned / 16)
                await metricsRec.recordFirstRange(Date().timeIntervalSince(tStart) * 1000.0)
                let tDec = Date()
                let plain = try AESCTRDecryptor.decrypt(
                    ciphertext: cipher, key: boot.key, baseIv: boot.iv, blockIndex: counterStart)
                await metricsRec.recordFirstDecrypt(Date().timeIntervalSince(tDec) * 1000.0)
                let wantedLen = Int(range.upperBound - range.lowerBound)
                let from = skip
                let to = min(from + wantedLen, plain.count)
                return plain.subdata(in: from..<to)
            } catch OfflineError.streamRangeFailed(let s) where s == 410 || s == 403 {
                throw OfflineError.streamRangeFailed(status: s)
            } catch OfflineError.sessionFingerprintMismatch {
                throw OfflineError.sessionFingerprintMismatch
            } catch {
                lastError = error
                let delay: UInt64 = [500, 1_000, 2_000][attempt] * 1_000_000
                try? await Task.sleep(nanoseconds: delay)
            }
        }
        throw OfflineError.streamNetworkExhausted(String(describing: lastError ?? URLError(.cannotConnectToHost)))
    }
}

// MARK: - Actor-based cache (Swift 6 concurrency safe)

private actor BootstrapCache {
    private var value: StreamSession.Bootstrap?
    func get() -> StreamSession.Bootstrap? { value }
    func set(_ b: StreamSession.Bootstrap) { value = b }
    func clear() { value = nil }
}

// MARK: - Metrics recorder

private actor MetricsRecorder {
    var bootstrapMs: Double?
    var firstRangeMs: Double?
    var firstDecryptMs: Double?
    var fileSizeBytes: Int64?
    var sessionRotations: Int = 0

    func recordBootstrap(_ ms: Double) { if bootstrapMs == nil { bootstrapMs = ms } }
    func recordFirstRange(_ ms: Double) { if firstRangeMs == nil { firstRangeMs = ms } }
    func recordFirstDecrypt(_ ms: Double) { if firstDecryptMs == nil { firstDecryptMs = ms } }
    func recordFileSize(_ n: Int64) { fileSizeBytes = n }
    func incRotation() { sessionRotations += 1 }
    func snapshot() -> StreamSession.SessionMetrics {
        StreamSession.SessionMetrics(
            bootstrapMs: bootstrapMs, firstRangeMs: firstRangeMs,
            firstDecryptMs: firstDecryptMs, fileSizeBytes: fileSizeBytes,
            sessionRotations: sessionRotations)
    }
}
